### ISSUE TITLE
Running on multiple GPUs

### ISSUE THREAD

**Issue by timwhittaker** [neuralgcm/neuralgcm#270]
I noticed that in Dinosaur, there's a way to shard using coordinates (as discussed in this [issue](https://github.com/neuralgcm/dinosaur/issues/45)). I was wondering if there’s a similar approach or functionality for sharding with coordinates in NeuralGCM.

I thought I could simple change the spmd_mesh attribute of the coords directly in the model:
```
mesh_devices = mesh_utils.create_device_mesh((2, 1, 1))
mesh = jax.sharding.Mesh(mesh_devices, ['z', 'x', 'y'])
model = neuralgcm.PressureLevelModel.from_checkpoint(ckpt)

updated_coords = replace(model._structure.coords, spmd_mesh=mesh)

# Create a new WhirlModel instance with updated coords
model._structure = model_builder.WhirlModel(
    coords=updated_coords,
    dt=model._structure.specs.dt,
    physics_specs=model._structure.specs.physics_specs,
    aux_features=model._structure.specs.aux_features,
    input_coords=model._structure.data_coords,
    output_coords=model._structure.data_coords,
    model_cls=model._structure.model_cls,
    to_xarray_fn=model._structure.to_xarray_fn,
    from_xarray_fn=model._structure.from_xarray_fn,
)
```

However, this approach is leading to shape mismatches during the encoding step.
Could you provide any guidance or examples on how I can apply this in NeuralGCM? Any insights would be greatly appreciated!

Thanks in advance!

**Comment by shoyer** [comment]
This is a great question. It's definitely possible (we do this in model training), although we never bothered to set this up for model inference. In our experience, it was not worth the trouble. Even the slowest NeuralGCM models (at 0.7 degrees) can simulate 20 years of climate per day on one TPU core -- and this model is not stable enough to simulate beyond weather timescales, anyways.

Is there a particular use-case that you had in mind here? You will definitely get higher throughput using multiple GPUs for data parallelism rather than model parallelism.

**Comment by timwhittaker** [comment]
Thanks for the response! 

My main goal is to use multiple GPUs because I’m computing gradients with respect to initial conditions. I managed to get this working for the 1.4-degree model on an A100 using checkpointing and chunking, but I’m running into issues scaling it to the 0.7-degree model. That’s what led me to explore sharding as a potential solution. 

Since I'm working with only one initial condition, I'm not sure how I could implement data parallelism. I’m fairly new to JAX, so it’s possible that I don’t fully understand how this process works yet, but from what I understand, in this context, data parallelism means splitting a dataset across multiple devices, where each device computes gradients for its subset of the data?
